ts/nni_manager/core/experimentsManager.ts

```diff
@@ -0,0 +1,160 @@
+        this.experimentsPath = newPath;
+    }
+
+    public setExperimentInfo(experimentId: string, key: string, value: any): void {
+        try {
+            if (this.profileUpdateTimer[key] !== undefined) {
+                clearTimeout(this.profileUpdateTimer[key]);
+                this.profileUpdateTimer[key] = undefined;
+            }
+            this.withLockSync(() => {
+                const experimentsInformation = JSON.parse(fs.readFileSync(this.experimentsPath).toString());
+                if (experimentsInformation[experimentId]) {
+                    experimentsInformation[experimentId][key] = value;
+                    fs.writeFileSync(this.experimentsPath, JSON.stringify(experimentsInformation));
+                } else {
+                    this.log.error(`Experiment Manager: Experiment Id ${experimentId} not found, this should not happen`);
+                }
+            });
+        } catch (err) {
+            this.log.error(err);
+            this.log.debug(`Experiment Manager: Retry set key value: ${experimentId} {${key}: ${value}}`);
+            this.profileUpdateTimer[key] = setTimeout(this.setExperimentInfo.bind(this), 100, experimentId, key, value);
+        }
+    }
+
+    private async withLock (func: Function, ...args: any): Promise<any> {
+        return withLock(func.bind(this), this.experimentsPath, {stale: 2 * 1000, retries: 100, retryWait: 100}, ...args);
+    }
+
+    private withLockSync (func: Function, ...args: any): any {
+        return withLockSync(func.bind(this), this.experimentsPath, {stale: 2 * 1000}, ...args);
+    }
+
+    private async readExperimentsInfo(): Promise<FileInfo> {
+        return this.withLock((path: string) => {
+                const buffer: Buffer = fs.readFileSync(path);
+                const mtime: number = fs.statSync(path).mtimeMs;
+                return {buffer: buffer, mtime: mtime};
+            },
+            this.experimentsPath);
+    }
+
+    private async checkCrashed(expId: string, pid: number): Promise<CrashedInfo> {
+        const alive: boolean = await isAlive(pid);
+        return {experimentId: expId, isCrashed: !alive}
+    }
+
+    private updateAllStatus(updateList: Array<string>, timestamp: number): {[key: string]: any} | undefined {
+        if (timestamp !== fs.statSync(this.experimentsPath).mtimeMs) {
+            return;
+        } else {
+            const experimentsInformation = JSON.parse(fs.readFileSync(this.experimentsPath).toString());
+            updateList.forEach((expId: string) => {
+                if (experimentsInformation[expId]) {
+                    experimentsInformation[expId]['status'] = 'STOPPED';
+                } else {
+                    this.log.error(`Experiment Manager: Experiment Id ${expId} not found, this should not happen`);
+                }
+            });
+            fs.writeFileSync(this.experimentsPath, JSON.stringify(experimentsInformation));
```

Comment:
This updateAllStatus() function logic does not need lock?

Reply:
Due to the cleanliness of the code and it is actually a sync function, lock it at line 46. If it is necessary to unify the lock position of the function, I will transferred the lock in `readExperimentsInfo()` in line 98.